### ISSUE TITLE
ctrl-c while donwloading packages destroys /dev/loop* on the host

### ISSUE THREAD

**Issue by gdamjan** [systemd/mkosi#107]
```
:: Proceed with installation? [Y/n] 
:: Retrieving packages...
 linux-api-headers-4.10.1-1-x86_64                                            852.4 KiB  64.9K/s 00:13 [############################################################] 100%
 tzdata-2017b-1-any                                                           235.8 KiB  44.3K/s 00:05 [############################################################] 100%
 iana-etc-20170512-1-any                                                      360.9 KiB  47.5K/s 00:08 [############################################################] 100%
 filesystem-2017.03-2-x86_64                                                   10.2 KiB  16.7K/s 00:01 [############################################################] 100%
^C‣ Unmounting Package Cache...                                               540.7 KiB   199K/s 00:40 [###---------------------------------------------------------]   6%

Interrupt signal received

‣ Unmounting Package Cache complete.
‣ Unmounting image...
‣ Unmounting image complete.
‣ Detaching image file...
‣ Detaching image file complete.
Traceback (most recent call last):
  File "/usr/bin/mkosi", line 2764, in <module>
    main()
  File "/usr/bin/mkosi", line 2760, in main
    build_stuff(args)
  File "/usr/bin/mkosi", line 2712, in build_stuff
    raw, tar, root_hash = build_image(args, workspace, run_build_script=False)
  File "/usr/bin/mkosi", line 2567, in build_image
    install_distribution(args, workspace.name, run_build_script, cached)
  File "/usr/bin/mkosi", line 1090, in install_distribution
    install[args.distribution](args, workspace, run_build_script)
  File "/usr/lib/python3.6/contextlib.py", line 53, in inner
    return func(*args, **kwds)
  File "/usr/bin/mkosi", line 988, in install_arch
    subprocess.run(cmdline, check=True)
  File "/usr/lib/python3.6/subprocess.py", line 405, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
  File "/usr/lib/python3.6/subprocess.py", line 828, in communicate
    self.wait()
  File "/usr/lib/python3.6/subprocess.py", line 1439, in wait
    (pid, sts) = self._try_wait(0)
  File "/usr/lib/python3.6/subprocess.py", line 1386, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
KeyboardInterrupt
Traceback (most recent call last):
  File "/usr/lib/python3.6/weakref.py", line 624, in _exitfunc
    f()
  File "/usr/lib/python3.6/weakref.py", line 548, in __call__
    return info.func(*info.args, **(info.kwargs or {}))
  File "/usr/lib/python3.6/tempfile.py", line 797, in _cleanup
    _shutil.rmtree(name)
  File "/usr/lib/python3.6/shutil.py", line 480, in rmtree
    _rmtree_safe_fd(fd, path, onerror)
  File "/usr/lib/python3.6/shutil.py", line 418, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  File "/usr/lib/python3.6/shutil.py", line 418, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  File "/usr/lib/python3.6/shutil.py", line 422, in _rmtree_safe_fd
    onerror(os.rmdir, fullname, sys.exc_info())
  File "/usr/lib/python3.6/shutil.py", line 420, in _rmtree_safe_fd
    os.rmdir(name, dir_fd=topfd)
OSError: [Errno 16] Device or resource busy: 'hugepages'
```
the host lost all of the `/dev/loop*` device nodes
```
$ ls -l /dev/loop-control
ls: cannot access '/dev/loop-control': No such file or directory
$ sudo rmmod loop
$ sudo modprobe loop
$ ls -l /dev/loop-control
crw-rw---- 1 root disk 10, 237 Jun 30 22:53 /dev/loop-control
```

**Comment by gdamjan** [comment]
`/dev/net/tun` and `/dev/kvm` also disappeared.

**Comment by poettering** [comment]
Uh oh. That smells more like an nspawn issue... any idea you can trace that with strace -f, to figure out where this is removed?

(I assume that nspawn is the culprit, since mkosi knows nothing of /dev/net/tun, but nspawn does)

**Comment by gdamjan** [comment]
there were other nodes that vanished. /dev/snd/ /dev/dm-1 and /dev/dm-2 but not /dev/dm-0.

I'll try to test it more, but it makes the system non-shutdown-able

**Comment by gdamjan** [comment]
diff between the output of `find /dev/ | sort`, before running mkosi and after breaking out of it - show a lot of deleted nodes:
```
--- /tmp/1.txt  2017-07-03 16:31:00.558284493 +0200
+++ /tmp/3.txt  2017-07-03 16:33:57.348089408 +0200
@@ -9,7 +9,16 @@
 /dev/block/259:2
 /dev/block/259:3
 /dev/block/259:4
-/dev/btrfs-control
+/dev/block/259:5
+/dev/block/259:6
+/dev/block/7:0
+/dev/block/7:1
+/dev/block/7:2
+/dev/block/7:3
+/dev/block/7:4
+/dev/block/7:5
+/dev/block/7:6
+/dev/block/7:7
 /dev/bus
 /dev/bus/usb
 /dev/bus/usb/001
@@ -20,7 +29,6 @@
 /dev/bus/usb/001/005
 /dev/bus/usb/002
 /dev/bus/usb/002/001
-/dev/cdc-wdm0
 /dev/char
 /dev/char/10:1
 /dev/char/10:130
@@ -32,6 +40,7 @@
 /dev/char/10:232
 /dev/char/10:235
 /dev/char/10:236
+/dev/char/10:237
 /dev/char/10:58
 /dev/char/10:59
 /dev/char/10:60
@@ -190,7 +199,6 @@
 /dev/cpu
 /dev/cpu_dma_latency
 /dev/cpu/microcode
-/dev/cuse
 /dev/disk
 /dev/disk/by-id
 /dev/disk/by-id/dm-name-home
@@ -210,14 +218,21 @@
 /dev/disk/by-id/nvme-SAMSUNG_MZVLW512HMJP-000L7_S359NX0HC18682-part3
 /dev/disk/by-id/nvme-SAMSUNG_MZVLW512HMJP-000L7_S359NX0HC18682-part4
 /dev/disk/by-label
+/dev/disk/by-label/EFI
 /dev/disk/by-label/ESP
 /dev/disk/by-label/HOME
+/dev/disk/by-label/root
 /dev/disk/by-label/ROOT
+/dev/disk/by-partlabel
+/dev/disk/by-partlabel/ESP\x20System\x20Partition
+/dev/disk/by-partlabel/Root\x20Partition
 /dev/disk/by-partuuid
 /dev/disk/by-partuuid/3fa90057-82f4-4f00-8700-abd52e3e73e7
 /dev/disk/by-partuuid/48ba020c-abe1-4ddd-a348-67b2df71e7cc
 /dev/disk/by-partuuid/7fd62259-5d63-4b60-b8ad-cf87174c12e2
+/dev/disk/by-partuuid/ab6c7e10-e9d5-427f-b763-14f98ec135f3
 /dev/disk/by-partuuid/b0a0807d-0592-40e9-adac-3bb724e9e305
+/dev/disk/by-partuuid/d95d222c-5ea8-4d6f-8519-94c1bdcc02b6
 /dev/disk/by-path
 /dev/disk/by-path/pci-0000:05:00.0-nvme-1
 /dev/disk/by-path/pci-0000:05:00.0-nvme-1-part1
@@ -226,76 +241,29 @@
 /dev/disk/by-path/pci-0000:05:00.0-nvme-1-part4
 /dev/disk/by-uuid
 /dev/disk/by-uuid/386cd0b9-abe6-4716-b412-6e3610091aef
+/dev/disk/by-uuid/4043f89a-d489-4369-b46b-47dd61852d94
+/dev/disk/by-uuid/553A-AE86
 /dev/disk/by-uuid/7b411dfd-9d7f-48bb-95fa-1ca52d67de57
 /dev/disk/by-uuid/ac26c4ed-3980-4cc7-8393-6a9d3badd3b5
 /dev/disk/by-uuid/af4ce52c-d63d-44a1-9bb7-9dab3f8694f6
 /dev/disk/by-uuid/c34f200e-5b9e-4e74-97a7-8a61ea563e75
 /dev/disk/by-uuid/F977-5DCE
 /dev/dm-0
-/dev/dm-1
-/dev/dm-2
-/dev/dri
-/dev/dri/by-path
-/dev/dri/by-path/pci-0000:00:02.0-card
-/dev/dri/by-path/pci-0000:00:02.0-render
-/dev/dri/card0
-/dev/dri/renderD128
-/dev/drm_dp_aux0
-/dev/drm_dp_aux1
-/dev/drm_dp_aux2
-/dev/fb0
 /dev/fd
 /dev/full
-/dev/fuse
-/dev/gpt-auto-root
-/dev/gpt-auto-root-luks
 /dev/hpet
 /dev/hugepages
 /dev/initctl
-/dev/input
-/dev/input/by-id
-/dev/input/by-id/usb-Chicony_Electronics_Co._Ltd._Integrated_Camera-event-if00
-/dev/input/by-path
-/dev/input/by-path/pci-0000:00:14.0-usb-0:8:1.0-event
-/dev/input/by-path/platform-pcspkr-event-spkr
-/dev/input/by-path/platform-thinkpad_acpi-event
-/dev/input/event0
-/dev/input/event1
-/dev/input/event10
-/dev/input/event11
-/dev/input/event12
-/dev/input/event13
-/dev/input/event14
-/dev/input/event15
-/dev/input/event16
-/dev/input/event17
-/dev/input/event2
-/dev/input/event3
-/dev/input/event4
-/dev/input/event5
-/dev/input/event6
-/dev/input/event7
-/dev/input/event8
-/dev/input/event9
-/dev/input/mice
-/dev/input/mouse0
-/dev/input/mouse1
 /dev/kmsg
-/dev/kvm
 /dev/log
-/dev/loop-control
 /dev/mapper
 /dev/mapper/control
 /dev/mapper/home
 /dev/mapper/root
 /dev/mapper/swap
-/dev/media0
-/dev/mei0
 /dev/mem
 /dev/memory_bandwidth
 /dev/mqueue
-/dev/net
-/dev/net/tun
 /dev/network_latency
 /dev/network_throughput
 /dev/null
@@ -305,12 +273,8 @@
 /dev/nvme0n1p2
 /dev/nvme0n1p3
 /dev/nvme0n1p4
-/dev/nvram
 /dev/port
-/dev/ppp
-/dev/psaux
 /dev/ptmx
-/dev/ptp0
 /dev/pts
 /dev/pts/0
 /dev/pts/1
@@ -318,31 +282,13 @@
 /dev/pts/3
 /dev/pts/ptmx
 /dev/random
-/dev/rfkill
 /dev/rtc
 /dev/rtc0
 /dev/shm
 /dev/snapshot
-/dev/snd
-/dev/snd/by-path
-/dev/snd/by-path/pci-0000:00:1f.3
-/dev/snd/controlC0
-/dev/snd/hwC0D0
-/dev/snd/hwC0D2
-/dev/snd/pcmC0D0c
-/dev/snd/pcmC0D0p
-/dev/snd/pcmC0D10p
-/dev/snd/pcmC0D3p
-/dev/snd/pcmC0D7p
-/dev/snd/pcmC0D8p
-/dev/snd/pcmC0D9p
-/dev/snd/seq
-/dev/snd/timer
 /dev/stderr
 /dev/stdin
 /dev/stdout
-/dev/tpm0
-/dev/tpmrm0
 /dev/tty
 /dev/tty0
 /dev/tty1
@@ -412,35 +358,10 @@
 /dev/ttyS1
 /dev/ttyS2
 /dev/ttyS3
-/dev/uhid
-/dev/uinput
 /dev/urandom
-/dev/v4l
-/dev/v4l/by-id
-/dev/v4l/by-id/usb-Chicony_Electronics_Co._Ltd._Integrated_Camera-video-index0
-/dev/v4l/by-path
-/dev/v4l/by-path/pci-0000:00:14.0-usb-0:8:1.0-video-index0
-/dev/v4l-touch0
 /dev/vcs
 /dev/vcs1
-/dev/vcs2
-/dev/vcs3
-/dev/vcs4
-/dev/vcs5
-/dev/vcs6
 /dev/vcsa
 /dev/vcsa1
-/dev/vcsa2
-/dev/vcsa3
-/dev/vcsa4
-/dev/vcsa5
-/dev/vcsa6
-/dev/vfio
-/dev/vfio/vfio
 /dev/vga_arbiter
-/dev/vhci
-/dev/vhost-net
-/dev/video0
-/dev/watchdog
-/dev/watchdog0
 /dev/zero
```

**Comment by gdamjan** [comment]
the strace log shows that the unlinkat calls were made by the mkosi process. let me try something

**Comment by gdiscry** [comment]
The second traceback looks like the cleanup of a `tempfile.TemporaryDirectory` (probably the workspace) failing when it tries to delete a directory called `hugepages` (like `/dev/hugepages`?).

After looking at `pacstrap`, I noticed that it uses `trap 'chroot_teardown' EXIT` to unmount `/dev` on exit.

I think that when the `KeyboardInterrupt` happens while `pacstrap` is running, the latter is killed (with `KILL`) and the teardown is not run. The workspace is then cleaned up with `/dev` still mounted.

I've tested a toy example with a python script calling a shell script and I can reproduce this behavior.
